tidysic/settings/formatted_string.py

```diff
@@ -0,0 +1,71 @@
+import re
+
+from tidysic.file.taggable import Taggable
+
+
+class FormattedString:
+    def __init__(self, raw_string: str):
+        try:
+            self.validate(raw_string)
+        except ValueError as e:
+            raise ValueError(f"Could not create formatted string, {e}")
+
+        self.raw_string = raw_string
+
+    def write(self, taggable: Taggable):
+        """
+        Produces the string built using the tags found in the given taggable
+        """
+        pattern = r"\{(.*?\{(\w+)(:.+?)?\}.*?)\}"
+        matches = re.findall(pattern, self.raw_string)
+
+        return_string = self.raw_string
+
+        substitutions = []
+        for to_substitute, tag_name, format_spec in matches:
+
+            value = getattr(taggable, tag_name, None)
+            if tag_name in Taggable.get_numeric_tag_names() and value:
+                value = int(value)
+            if tag_name in Taggable.get_non_numeric_tag_names() and not value:
+                value = f"Unknown {tag_name}"
+
+            formattable = to_substitute.replace(
+                f"{{{tag_name}{format_spec}}}", f"{{{format_spec}}}"
+            )
+            substitutions.append(
+                (f"{{{to_substitute}}}", formattable.format(value) if value else "")
+            )
+
+        for old, new in substitutions:
```

Comment:
I didn't look deeply into it, but couldn't we somehow use `re.substitute(pattern, substitution)` directly instead of building a list of substitution and then replacing?

Reply:
You're right. Building a list of substitutions doesn't make any sense, I should apply them right away.

These are exact text match, `str.replace` should do the trick.

Reply:
While modifying this, I also switch the string formatting method from f-string to the `"%s" % value` method. This helps avoid using double brackets to escape a single bracket, helping make the process clearer (or so I hope).